Bluetooth: Controller: Single recv thread for HCI-only builds

cvinayak opened the pull request:
Updated implementation to use single receive thread to enqueue HCI ISO data, ACL data and events towards Host when building HCI-Only samples/applications, i.e. when building hci_uart, hci_spi or hci_ipc samples (CONFIG_BT_HCI_RAW=y).

This implementation will serialize HCI events and data as they occur corresponding to on-air timelines of their occurrences which is how they are generated by the Link Layer.

cvinayak committed:
Bluetooth: Controller: Single recv thread for HCI-only builds

Updated implementation to use single receive thread to
enqueue HCI ISO data, ACL data and events towards Host when
building HCI-Only samples/applications, i.e. when building
hci_uart, hci_spi or hci_ipc samples (CONFIG_BT_HCI_RAW=y).

This implementation will serialize HCI events and data as
they occur corresponding to on-air timelines of their
occurrences which is how they are generated by the Link
Layer.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@nashif, any idea why @Tronil  does not get added as reviewer (MAINTAINERS.yml file has his entry)?.... and @ppryga-nordic  probably does not get added as his github id does not match?, I will need to send a PR to fix it.

@ppryga-nordic should I remove your entry or do you want it corrected?

Thalley commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -425,12 +413,104 @@ static void prio_recv_thread(void *p1, void *p2, void *p3)
+		/* There may still be completed nodes, continue pushing all those up to Host before
+		 * waiting/polling for sem_recv.
+		 */
+	} while (iso_received || node_rx);
```
Since `node_rx` isn't used after this, you could consider renaming `iso_received` to `evt_received` and set that to true if there's ISO data or an node_rx, which would make this check just be `while(evt_received)`

Thalley commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -676,44 +757,84 @@ static void recv_thread(void *p1, void *p2, void *p3)
 	const struct device *dev = p1;
 	struct hci_driver_data *data = dev->data;
 
+#if defined(CONFIG_BT_HCI_ACL_FLOW_CONTROL) || !defined(CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE)
+	enum {
 #if defined(CONFIG_BT_HCI_ACL_FLOW_CONTROL)
+		EVENT_SIGNAL,
+#endif /* CONFIG_BT_HCI_ACL_FLOW_CONTROL */
+
+#if !defined(CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE)
+		EVENT_SEM,
+#endif /* !CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE */
+
+		EVENT_FIFO,
+		EVENT_MAX,
+	};
```
What do we gain from guarding the enum and/or the values in the enum?

Thalley replied:
If we don't guard them, it looks like we can use `IS_ENABLED` more below

Thalley commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -789,7 +926,15 @@ static int iso_handle(const struct device *dev, struct net_buf *buf)
 	if (evt) {
 		LOG_DBG("Replying with event of %u bytes", evt->len);
+
+#if defined(CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE)
 		bt_recv_prio(dev, evt);
+
+#else /* CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE */
+		struct hci_driver_data *data = dev->data;
+
+		data->recv(dev, evt);
+#endif /* CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE */
```
We could consider a helper function for these lines as they are copied several times in this file

Thalley commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -676,44 +757,84 @@ static void recv_thread(void *p1, void *p2, void *p3)
 
-		events[0].state = K_POLL_STATE_NOT_READY;
-		events[1].state = K_POLL_STATE_NOT_READY;
+		for (int i = 0; i < ARRAY_SIZE(events); i++) {
```
Oh, wasn't aware of that macro :)

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -116,10 +116,27 @@ config BT_CTLR_TIFS_HW_SUPPORT
 
+config BT_CTLR_RX_POLL
+	# Hidden, Enable Async I/O Framework, i.e. k_poll
+	bool
+	depends on BT_HCI_RAW || BT_HCI_ACL_FLOW_CONTROL
+	select POLL
+	default y
+
 config BT_CTLR_RX_PRIO_STACK_SIZE
-	# Hidden, Controller's Co-Operative high priority Rx thread stack size.
+	# Hidden
```
I've seen other reviewers requesting that we don't put e.g. `# hidden` (and even some who have removed them later). 

I'm OK with it as is, but just a note :)

cvinayak replied:
This is a comment that makes it obvious of a hidden Kconfig when reading through this file when it appears similar to other non-hidden Kconfig that have `depends on`, `default` and `help`.

Thalley commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -425,12 +420,114 @@ static void prio_recv_thread(void *p1, void *p2, void *p3)
+#else /* !CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE */
+static void node_rx_recv(const struct device *dev)
+{
+#if defined(CONFIG_BT_CONN) || defined(CONFIG_BT_CTLR_ADV_ISO)
+	const struct hci_driver_data *data = dev->data;
+#endif /* CONFIG_BT_CONN || CONFIG_BT_CTLR_ADV_ISO */
```
Not quite sure I understand your last comment here. Looks to me that this can be moved into the 
```
		while (num_cmplt != 0U) {
#if defined(CONFIG_BT_CONN) || defined(CONFIG_BT_CTLR_ADV_ISO)
```
part to reduce scope

cvinayak replied:
`dev->data->recv()`  .... this is double dereferencing, two indirect address be loaded to call a function.

Keeping one outside the loop will need single indirect address loaded to call the function (similar to a function pointer that it is already) in the loop.

Thalley replied:
Ah, so if I understand you correct, this is done to not dereference `dev` for each loop iteration? 

I was under the impression that the compiler would not do that, especially not when the expression is `const`, but I guess that since `dev->data` can in theory be changed per iteration from another thread, it cannot do that.

cvinayak replied:
>if I understand you correct, this is done to not dereference dev for each loop iteration?

Yes.  In the `struct device` `*data` is not const. Though in this content `data` or `*data` is not being changed, the compiler does not know that, hence we make it obvious for the compiler by using `const struct hci_driver_data *data = dev->data;`

Thalley commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -789,7 +938,12 @@ static int iso_handle(const struct device *dev, struct net_buf *buf)
 		LOG_DBG("Replying with event of %u bytes", evt->len);
-		bt_recv_prio(dev, evt);
+
+#if defined(CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE)
+		err = bt_recv_prio(dev, evt);
+#else /* CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE */
+		err = bt_recv(dev, evt);
+#endif /* CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE */
```
Not neccesarrily something that needs an update, but `bt_recv` is only ever used in this combination. We could have `bt_recv` call `bt_recv_prio` depending on `CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE` so that this just becomes
```suggestion
		err = bt_recv(dev, evt);
```

Semantically that's not a good solution though I admit, given the naming of these functions

cvinayak replied:
`bt_recv` and `bt_recv_prio`  is more clear as to the priority of the function execution, lets keep it as-is.

Thalley reviewed (dismissed):
No blockers

PavelVPV reviewed (dismissed):
I confirm that this fixes an issue observed when running `hfc` bsim test on nRF5340 where incoming ACL data was incorrectly passed to Host after disconnection thus causing `No available ACL buffers!` error in `hci_ipc` driver.

The `hfc` test will be enabled in #87815 to prevent regression.

cvinayak commented:
Reverted, and added back the deleted `hci_disconn_complete_process()`, and calling it now from `hci_disconn_complete_encode()` when Rx priority thread is not used.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -8508,6 +8508,10 @@ void hci_disconn_complete_encode(struct pdu_data *pdu_data, uint16_t handle,
 	ep->handle = sys_cpu_to_le16(handle);
 	ep->reason = *((uint8_t *)pdu_data);
+
+#if !defined(CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE)
+	hci_disconn_complete_process(handle);
+#endif /* CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE */
```
This looks like the incorrect place to process the event. The function is called `_encode`, so sounds like this function should be called near where `hci_disconn_complete_encode` is called instead.
